x/ccv/consumer/keeper/keeper.go

```diff
@@ -206,50 +206,95 @@ func (k Keeper) DeletePendingChanges(ctx sdk.Context) {
-func (k Keeper) IteratePacketMaturityTime(ctx sdk.Context, cb func(vscId, timeNs uint64) (stop bool)) {
+// GetVSCPacketQueueTimeSlice gets a specific VSC packet queue timeslice.
+// A timeslice is a slice of IDs corresponding to received VSC packets
+// that reach maturity at a certain time.
+func (k Keeper) GetVSCPacketQueueTimeSlice(ctx sdk.Context, timestamp time.Time) (vscIDs []uint64) {
 	store := ctx.KVStore(k.storeKey)
-	iterator := sdk.KVStorePrefixIterator(store, []byte{types.PacketMaturityTimeBytePrefix})
 
-	defer iterator.Close()
-	for ; iterator.Valid(); iterator.Next() {
-		// Extract bytes following the 1 byte prefix
-		seqBytes := iterator.Key()[1:]
-		seq := binary.BigEndian.Uint64(seqBytes)
+	bz := store.Get(types.VSCPacketQueueKey(timestamp))
+	if bz == nil {
+		return []uint64{}
+	}
 
-		timeNs := binary.BigEndian.Uint64(iterator.Value())
+	validatorSetChangeIDs := types.ValidatorSetChangeIDs{}
+	k.cdc.MustUnmarshal(bz, &validatorSetChangeIDs)
 
-		stop := cb(seq, timeNs)
-		if stop {
-			break
-		}
+	return validatorSetChangeIDs.Ids
+}
+
+// SetVSCPacketQueueTimeSlice sets a specific VSC packet queue timeslice.
+func (k Keeper) SetVSCPacketQueueTimeSlice(ctx sdk.Context, timestamp time.Time, vscIDs ...uint64) {
+	store := ctx.KVStore(k.storeKey)
+	bz := k.cdc.MustMarshal(&types.ValidatorSetChangeIDs{Ids: vscIDs})
+	store.Set(types.VSCPacketQueueKey(timestamp), bz)
+}
+
+// InsertVSCPacketQueue inserts the ID of a VSC packet to the appropriate timeslice
+// in the VSCPacketQueue.
+func (k Keeper) InsertVSCPacketQueue(ctx sdk.Context, vscID uint64, maturityTime time.Time) {
+	timeSlice := k.GetVSCPacketQueueTimeSlice(ctx, maturityTime)
+	if len(timeSlice) == 0 {
+		k.SetVSCPacketQueueTimeSlice(ctx, maturityTime, vscID)
+	} else {
+		timeSlice = append(timeSlice, vscID)
+		k.SetVSCPacketQueueTimeSlice(ctx, maturityTime, timeSlice...)
 	}
 }
 
-// SetPacketMaturityTime sets the maturity time for a given received VSC packet id
-func (k Keeper) SetPacketMaturityTime(ctx sdk.Context, vscId, maturityTime uint64) {
+// VSCPacketQueueIterator returns all the VSC packet queue timeslices from time 0 until endTime.
+func (k Keeper) VSCPacketQueueIterator(ctx sdk.Context, endTime time.Time) sdk.Iterator {
 	store := ctx.KVStore(k.storeKey)
-	timeBytes := make([]byte, 8)
-	binary.BigEndian.PutUint64(timeBytes, maturityTime)
-	store.Set(types.PacketMaturityTimeKey(vscId), timeBytes)
+	return store.Iterator([]byte{types.VSCPacketQueueBytePrefix},
+		sdk.InclusiveEndBytes(types.VSCPacketQueueKey(endTime)))
 }
 
-// GetPacketMaturityTime gets the maturity time for a given received VSC packet id
-func (k Keeper) GetPacketMaturityTime(ctx sdk.Context, vscId uint64) uint64 {
+// DequeueAllMatureVSCPacketQueue returns a concatenated list of all IDs of matured VSC packets
+// and deletes the timeslices from the VSC packet queue.
+func (k Keeper) DequeueAllMatureVSCPacketQueue(ctx sdk.Context) (vscIDs []uint64) {
```

Comment:
This deletes iterator keys during iteration.

From the rest of our codebase it seems that we used to prefer explicit `DeleteX` calls inside functions, ie:
```golang
func HandleItems() {
  items := GetItems()
  doAction(items)
  DeleteItems(items...)
}

```

Are we attempting to move from the previous pattern?

Reply:
It's just the code from the SDK. But I agree that we should keep to our pattern.